File: web/src/forms/rules.ts
/**
 * Conditional requirements, one third of a three-platform contract (I5, SRS 7.2).
 *
 * The same subset of JSON Logic the backend evaluates in `app/forms/rules.py` and the phone
 * evaluates in `core:sync`. Agreement is not achieved by writing the algorithm carefully three
 * times — that produces three algorithms that match on the cases somebody thought of. It is
 * achieved by `forms/contract/validation-cases.json`, which all three test suites run.
 *
 * Two JavaScript-specific traps the corpus pins down, and both would be invisible in review:
 *
 * * **Falsy is not empty.** `if (!value)` treats `0` and `false` as blanks, so a measured 0 Ω
 *   and a "was it sign-posted? no" would both be reported as unanswered.
 * * **`==` coerces.** `11 == '11'` is true, so a rule would fire here and not on the server.
 *   Every comparison below is type-aware for that reason.
 */

/** A rule as the block library declares it. Values come from JSON, hence `unknown`. */
export interface FormRule {
  when?: unknown;
  require?: unknown;
  message?: unknown;
  blocks_execution?: boolean;
}

export interface MissingRequirement {
  field: string;
  /** The rule's own message when it has one. Presentation: the contract is the field. */
  message: string | null;
}

export type Answers = Record<string, unknown>;

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/**
 * Whether a value counts as an answer.
 *
 * Deliberately not `Boolean(value)`: zero is a measurement, `false` is a reply, and a field
 * holding a space reads as empty both on screen and on paper.
 */
export function isAnswered(value: unknown): boolean {
  if (value === null || value === undefined) return false;
  if (typeof value === 'boolean' || typeof value === 'number') return true;
  if (typeof value === 'string') return value.trim().length > 0;
  if (Array.isArray(value)) return value.length > 0;
  if (isPlainObject(value)) return Object.keys(value).length > 0;
  return true;
}

/** Resolve an operand: `{var: 'x'}` reads an answer, anything else is a literal. */
export function resolve(token: unknown, answers: Answers): unknown {
  if (isPlainObject(token) && 'var' in token) {
    const name = token.var;
    return typeof name === 'string' ? answers[name] : undefined;
  }
  return token;
}

/**
 * Equality that does not coerce across types.
 *
 * `11` and `'11'` are different values, and a boolean is never 0 or 1 — otherwise "was it
 * sign-posted?" would match a count.
 */
function same(left: unknown, right: unknown): boolean {
  if (typeof left !== typeof right) return false;
  return left === right;
}

function operandPair(operands: unknown): [unknown, unknown] | null {
  if (!Array.isArray(operands) || operands.length !== 2) return null;
  return [operands[0], operands[1]];
}

/** Evaluate one condition. Anything unrecognised is `false`, never a thrown error. */
export function evaluateCondition(condition: unknown, answers: Answers): boolean {
  if (!isPlainObject(condition) || Object.keys(condition).length === 0) return false;

  if ('==' in condition) {
    const pair = operandPair(condition['==']);
    return pair !== null && same(resolve(pair[0], answers), resolve(pair[1], answers));
  }

  if ('!=' in condition) {
    const pair = operandPair(condition['!=']);
    return pair !== null && !same(resolve(pair[0], answers), resolve(pair[1], answers));
  }

  if ('in' in condition) {
    const pair = operandPair(condition['in']);
    if (pair === null) return false;
    const needle = resolve(pair[0], answers);
    const haystack = resolve(pair[1], answers);
    if (typeof haystack === 'string') {
      // JSON Logic defines `in` over a string as substring containment.
      return typeof needle === 'string' && haystack.includes(needle);
    }
    if (Array.isArray(haystack)) return haystack.some((item) => same(needle, item));
    // Neither a list nor a string: false, not an exception. A badly written rule must not stop a
    // technician from sending a day's work.
    return false;
  }

  if ('and' in condition) {
    const parts = condition.and;
    if (!Array.isArray(parts) || parts.length === 0) return false;
    return parts.every((part) => evaluateCondition(part, answers));
  }

  if ('or' in condition) {
    const parts = condition.or;
    if (!Array.isArray(parts) || parts.length === 0) return false;
    return parts.some((part) => evaluateCondition(part, answers));
  }

  return false;
}

/**
 * Fields the rules demand and the answers do not supply, in the rules' own order.
 *
 * Deduplicated: two rules may demand the same field, and telling the crew twice about one empty
 * box is a defect of the platform, not a fact about their work. The order is the rules' order, so
 * the list a technician reads does not shuffle between the phone and the web.
 */
export function missingRequirements(rules: unknown, answers: Answers): MissingRequirement[] {
  if (!Array.isArray(rules)) return [];
  const found: MissingRequirement[] = [];
  const seen = new Set<string>();

  for (const rule of rules) {
    if (!isPlainObject(rule)) continue;
    const required = rule.require;
    if (!Array.isArray(required) || required.length === 0) continue;
    if (!evaluateCondition(rule.when, answers)) continue;
    const message = typeof rule.message === 'string' && rule.message ? rule.message : null;
    for (const field of required) {
      const key = String(field);
      if (seen.has(key) || isAnswered(answers[key])) continue;
      seen.add(key);
      found.push({ field: key, message });
    }
  }
  return found;
}

/** The messages a technician reads, which is what a screen shows. */
export function requirementMessages(rules: unknown, answers: Answers): string[] {
  return missingRequirements(rules, answers).map(
    (missing) => missing.message ?? `${missing.field}: es obligatorio en este caso`,
  );
}

File: backend/app/forms/rules.py
"""Conditional requirements, evaluated identically on three platforms (I5, SRS 7.2).

The blocks declare their rules in JSON Logic so the backend, the web and the phone agree. Writing
the same algorithm three times does not produce agreement; it produces three algorithms that
happen to match on the cases somebody thought of. What produces agreement is
``forms/contract/validation-cases.json``: one corpus, run by all three test suites, that fails
when an implementation drifts.

The failure that corpus prevents is expensive in the field, in both directions. If the phone
accepts a capture the server later rejects, the crew's work comes back the next day and nobody
knows why. If the phone demands a field the server does not, the crew cannot close the work order
with the network down — which is the situation the whole product exists for.

The evaluator is a **subset** of JSON Logic on purpose: only the operators the block library
uses. A full library would be more general, and would also accept expressions the mobile renderer
cannot evaluate. An unknown operator is False rather than an error, because a badly written rule
must not stop a technician from sending a day's work; it is caught by catalogue validation,
before the form ever reaches a phone.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class MissingRequirement:
    """One field a rule demanded and the answers do not supply."""

    field: str
    #: The rule's own message, when it has one. Presentation: the contract is the field.
    message: str | None = None


def is_answered(value: Any) -> bool:
    """Whether a value counts as an answer.

    Three cases worth naming, because each is a way all three implementations get it wrong at
    once:

    * **Zero is an answer.** A measured 0 Ω is a result, and one of the ones that matter.
    * **False is an answer.** "Was it sign-posted? No" is a reply, not a blank.
    * **Whitespace is not.** A field holding a space reads as empty on the screen and on paper.
    """
    if value is None:
        return False
    if isinstance(value, bool | int | float):
        return True
    if isinstance(value, str):
        return bool(value.strip())
    if isinstance(value, Mapping):
        return bool(value)
    if isinstance(value, Sequence):
        return bool(value)
    return True


def resolve(token: Any, answers: Mapping[str, Any]) -> Any:
    """Resolve an operand: ``{"var": "x"}`` reads an answer, anything else is a literal."""
    if isinstance(token, Mapping) and "var" in token:
        return answers.get(token["var"])
    return token


def _same(left: Any, right: Any) -> bool:
    """Equality that does not coerce across types.

    ``11`` and ``"11"`` are different values. Without this, a platform with loose comparison
    would fire rules the others do not — and booleans would compare equal to 0 and 1, which is
    how "was it sign-posted?" ends up matching a count.
    """
    if isinstance(left, bool) != isinstance(right, bool):
        return False
    if isinstance(left, str) != isinstance(right, str):
        return False
    return bool(left == right)


def evaluate_condition(condition: Any, answers: Mapping[str, Any]) -> bool:
    """Evaluate one JSON Logic condition. Anything unrecognised is False."""
    if not isinstance(condition, Mapping) or not condition:
        return False

    if "==" in condition:
        operands = condition["=="]
        if not isinstance(operands, Sequence) or len(operands) != 2:
            return False
        return _same(resolve(operands[0], answers), resolve(operands[1], answers))

    if "!=" in condition:
        operands = condition["!="]
        if not isinstance(operands, Sequence) or len(operands) != 2:
            return False
        return not _same(resolve(operands[0], answers), resolve(operands[1], answers))

    if "in" in condition:
        operands = condition["in"]
        if not isinstance(operands, Sequence) or len(operands) != 2:
            return False
        needle = resolve(operands[0], answers)
        haystack = resolve(operands[1], answers)
        if isinstance(haystack, str):
            # JSON Logic defines `in` over a string as substring containment.
            return isinstance(needle, str) and needle in haystack
        if isinstance(haystack, Sequence) and not isinstance(haystack, str | bytes):
            return any(_same(needle, item) for item in haystack)
        # Neither a list nor a string: False, not an exception.
        return False

    # «Está contestado» y su complemento. Escritos con `is_answered`, la misma función que decide
    # el otro lado de la regla: si un 0 cuenta como respuesta para `require`, tiene que contar
    # igual para la condición, o una regla diría una cosa y su exigencia otra.
    if "!!" in condition:
        return is_answered(resolve(condition["!!"], answers))

    if "!" in condition:
        return not is_answered(resolve(condition["!"], answers))

    if "and" in condition:
        parts = condition["and"]
        if not isinstance(parts, Sequence) or not parts:
            return False
        return all(evaluate_condition(part, answers) for part in parts)

    if "or" in condition:
        parts = condition["or"]
        if not isinstance(parts, Sequence) or not parts:
            return False
        return any(evaluate_condition(part, answers) for part in parts)

    return False


def missing_requirements(
    rules: Sequence[Mapping[str, Any]], answers: Mapping[str, Any]
) -> list[MissingRequirement]:
    """Fields the rules demand and the answers do not supply, in the rules' own order.

    In order, and deduplicated: two rules may demand the same field, and telling the crew twice
    about one empty box is a defect of the platform rather than a fact about their work. The
    order is the rules' order so the list a technician reads does not shuffle between platforms.
    """
    found: list[MissingRequirement] = []
    seen: set[str] = set()
    for rule in rules:
        if not isinstance(rule, Mapping):
            continue
        required = rule.get("require") or []
        if not required or not isinstance(required, Sequence) or isinstance(required, str):
            continue
        if not evaluate_condition(rule.get("when"), answers):
            continue
        message = rule.get("message")
        for field in required:
            key = str(field)
            if key in seen or is_answered(answers.get(key)):
                continue
            seen.add(key)
            found.append(MissingRequirement(field=key, message=str(message) if message else None))
    return found
